File: frontend/src/lib/validations/contrato.ts
/* ──────────────────────────────────────────────────────────────────────────
 * ITER TIC — Schema de validação Zod para Contrato
 * Espelha as regras do Pydantic no backend (ContratoCreate)
 *
 * Equipe de Fiscalização: cada papel aceita 1 titular_id + N substitutos_ids.
 * ────────────────────────────────────────────────────────────────────────── */

import { z } from "zod";



/* ── Sub-schema: Itens do contrato ────────────── */

const itemContratoSchema = z.object({
  quantidade: z
    .number({ error: "Quantidade é obrigatória." })
    .int("Deve ser inteiro.")
    .min(1, "Mínimo 1."),
  valor_unitario: z
    .number({ error: "Valor unitário é obrigatório." })
    .min(0, "Não pode ser negativo."),
  tipo_catalogo: z.enum(["CATMAT", "CATSER", ""]).optional().or(z.literal(null)),
  codigo_catalogo: z.string().optional().or(z.literal(null)),
  catalogo_produto_id: z.number({ error: "Item do catálogo é obrigatório." }).min(1, "Item do catálogo é obrigatório."),
  data_inicio_vigencia: z.string().optional().or(z.literal("")).nullable(),
  data_fim_vigencia: z.string().optional().or(z.literal("")).nullable(),
});

/* ── Sub-schema: um papel da equipe (titular + substitutos) ────────────── */

const equipePapelSchema = z.object({
  titulares_ids: z.array(z.number().int()).default([]),
  substitutos_ids: z.array(z.number().int()).default([]),
});

const equipeSchema = z.object({
  gestor: equipePapelSchema.default({ titulares_ids: [], substitutos_ids: [] }),
  fiscal_requisitante: equipePapelSchema.default({ titulares_ids: [], substitutos_ids: [] }),
  fiscal_tecnico: equipePapelSchema.default({ titulares_ids: [], substitutos_ids: [] }),
  fiscal_administrativo: equipePapelSchema.default({ titulares_ids: [], substitutos_ids: [] }),
});

/* ── Schema principal ──────────────────────────────────────────────────── */

export const contratoCreateSchema = z
  .object({
    /* Projeto de Origem */
    projeto_id: z
      .number({ error: "Selecione o projeto de origem." })
      .int()
      .min(1, "Selecione um projeto válido."),

    /* Modalidade da Contratação */
    modalidade_contrato: z.enum(["CONTRATO", "ARP"]).default("CONTRATO"),

    /* Tipo de Contratação */
    tipo_contratacao: z.string({ error: "Selecione o tipo de contratação." }).min(1, "Selecione o tipo de contratação."),

    /* Dados da Contratação */
    numero: z.coerce.number().int().positive("Número deve ser maior que 0."),
    ano: z.coerce.number().int().min(2015, "Ano inválido.").max(2035, "Ano inválido."),

    /* Tipo de Instrumento (obrigatório para CONTRATO) */
    tipo_instrumento: z.enum(["CONTRATO", "NOTA_EMPENHO"]).optional().nullable(),

    fornecedor_id: z
      .number()
      .int()
      .optional()
      .or(z.literal(0)),

    tipo_fornecedor_contrato: z
      .enum(["REVENDEDOR", "FABRICANTE", "REVENDEDOR_E_FABRICANTE", ""])
      .optional()
      .or(z.literal(null)),

    tipo_contrato: z.enum(
      ["Aquisição", "Serviço continuado", "Subscrição"],
      { error: "Selecione o tipo de contrato." }
    ),

    /* Itens da Contratação */
    itens: z
      .array(itemContratoSchema)
      .min(1, "Adicione pelo menos um item à contratação."),

    /* Vigência */
    data_inicio_vigencia: z
      .string()
      .optional()
      .or(z.literal("")),

    vigencia_meses: z
      .number()
      .int()
      .min(0)
      .optional()
      .nullable(),

    prorrogacao_meses: z
      .number()
      .int()
      .min(0)
      .max(120)
      .default(0),

    data_assinatura: z
      .string({ error: "Data de assinatura é obrigatória." })
      .min(1, "Data de assinatura é obrigatória."),

    data_fim_vigencia: z
      .string({ error: "Data de fim de vigência é obrigatória." })
      .min(1, "Data de fim é obrigatória."),

    situacao_atual: z.enum(
      ["Vigente", "Extinto", "Extinto, mas suporte vigente"],
      { error: "Selecione a situação." }
    ),


    /* Campo específico de ARP */
    orgao_gerenciador: z.string().max(300).optional().or(z.literal("")),

    /* Equipe de Fiscalização (novo formato: titular + substitutos) */
    equipe: equipeSchema.optional(),
  })
  .refine(
    (d) => {
      if (d.data_assinatura && d.data_fim_vigencia) {
        return d.data_fim_vigencia >= d.data_assinatura;
      }
      return true;
    },
    {
      message: "A data de fim de vigência deve ser posterior à data de assinatura.",
      path: ["data_fim_vigencia"],
    }
  )
  .refine(
    (d) => {
      // tipo_instrumento é obrigatório para modalidade CONTRATO
      if (d.modalidade_contrato === "CONTRATO") {
        return !!d.tipo_instrumento;
      }
      return true;
    },
    {
      message: "Selecione o tipo de instrumento.",
      path: ["tipo_instrumento"],
    }
  );

export type ContratoCreateFormData = z.infer<typeof contratoCreateSchema>;

/* ── Limpar payload antes de enviar para a API ─────────────────────────── */

export function cleanContratoPayload(
  data: ContratoCreateFormData
): Record<string, unknown> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const clean: Record<string, any> = { ...data };

  // Converter fornecedor_id = 0 → null
  if (clean.fornecedor_id === 0) clean.fornecedor_id = null;

  // Converter tipo_fornecedor_contrato vazio → null
  if (!clean.tipo_fornecedor_contrato) clean.tipo_fornecedor_contrato = null;

  // Limpar campos de catálogo vazios nos itens
  if (clean.itens && Array.isArray(clean.itens)) {
    clean.itens = clean.itens.map((item: any) => ({
      ...item,
      tipo_catalogo: item.tipo_catalogo === "" ? null : item.tipo_catalogo,
      codigo_catalogo: item.codigo_catalogo === "" ? null : item.codigo_catalogo,
      catalogo_produto_id: item.catalogo_produto_id === 0 || item.catalogo_produto_id === "" ? null : item.catalogo_produto_id,
      data_inicio_vigencia: item.data_inicio_vigencia === "" ? null : item.data_inicio_vigencia,
      data_fim_vigencia: item.data_fim_vigencia === "" ? null : item.data_fim_vigencia,
    }));
  }

  // Limpar strings vazias → null
  for (const key of ["data_inicio_vigencia"]) {
    if (clean[key] === "") clean[key] = null;
  }

  // orgao_gerenciador: limpar se CONTRATO ou vazio
  if (clean.modalidade_contrato !== "ARP" || !clean.orgao_gerenciador) {
    clean.orgao_gerenciador = null;
  }

  // tipo_instrumento: limpar se ARP
  if (clean.modalidade_contrato === "ARP") {
    clean.tipo_instrumento = null;
  }

  // vigencia_meses 0 ou undefined → null
  if (!clean.vigencia_meses) clean.vigencia_meses = null;

  // Transformar equipe: filtrar IDs inválidos
  if (clean.equipe) {
    const equipeLimpa: Record<string, unknown> = {};
    for (const papel of [
      "gestor",
      "fiscal_requisitante",
      "fiscal_tecnico",
      "fiscal_administrativo",
    ]) {
      const p = clean.equipe[papel];
      if (p) {
        equipeLimpa[papel] = {
          titulares_ids: (p.titulares_ids || []).filter(
            (id: number) => id > 0
          ),
          substitutos_ids: (p.substitutos_ids || []).filter(
            (id: number) => id > 0
          ),
        };
      }
    }
    clean.equipe = equipeLimpa;
  }

  return clean;
}

File: app/models/contrato.py
"""
ITER TIC - Modelos SQLAlchemy V2 do Módulo 3: Execução e Fiscalização (Contratos)

Arquitetura:
  - Um Contrato nasce de um Projeto com status 'Licitação concluída'.
  - Cada Contrato possui uma equipe de fiscalização (Gestor, Fiscal
    Requisitante, Fiscal Técnico, Fiscal Administrativo), referenciando
    a tabela Servidor.
  - Campos financeiros segregados em Investimento e Custeio conforme
    a natureza de despesa orçamentária.
"""

from __future__ import annotations

import enum
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import (
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  ENUMS DO MÓDULO 3                                                     ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class TipoContratoEnum(str, enum.Enum):
    """Classificação do tipo de contratação."""
    AQUISICAO = "Aquisição"
    SERVICO_CONTINUADO = "Serviço continuado"
    SUBSCRICAO = "Subscrição"


class ComplexidadeContratoEnum(str, enum.Enum):
    """Complexidade do contrato."""
    SIMPLES = "simples"
    INTERMEDIARIA = "intermediaria"
    COMPLEXA = "complexa"



class SituacaoContratoEnum(str, enum.Enum):
    """Situação atual do contrato na instituição."""
    VIGENTE = "Vigente"
    VIGENTE_SUSPENSO = "Vigente com execução suspensa"
    VIGENTE_PRORROGADO = "Vigente prorrogado"
    EXTINTO = "Extinto"
    EXTINTO_OBRIGACOES = "contrato extinto com obrigações remanescentes"


class ModalidadeContratoEnum(str, enum.Enum):
    """Modalidade da contratação."""
    CONTRATO = "CONTRATO"
    ARP = "ARP"


class TipoInstrumentoEnum(str, enum.Enum):
    """Tipo de instrumento contratual (apenas para modalidade CONTRATO)."""
    CONTRATO = "CONTRATO"
    NOTA_EMPENHO = "NOTA_EMPENHO"


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  CONTRATO (entidade central do Módulo 3)                               ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class PapelEquipeEnum(str, enum.Enum):
    """Papel de um servidor na equipe de fiscalização."""
    GESTOR = "gestor"
    FISCAL_REQUISITANTE = "fiscal_requisitante"
    FISCAL_TECNICO = "fiscal_tecnico"
    FISCAL_ADMINISTRATIVO = "fiscal_administrativo"


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  CONTRATO (entidade central do Módulo 3)                               ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class Contrato(Base):
    __tablename__ = "contratos"

    # ── Identificação ───────────────────────────────────────────────────────
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    numero: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    ano: Mapped[int] = mapped_column(Integer, nullable=False, index=True)

    # ── Modalidade ───────────────────────────────────────────────────────────
    modalidade_contrato: Mapped[ModalidadeContratoEnum] = mapped_column(
        Enum(
            ModalidadeContratoEnum,
            name="modalidade_contrato_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
        default=ModalidadeContratoEnum.CONTRATO,
        server_default="CONTRATO",
        index=True,
        comment="Modalidade: CONTRATO ou ARP.",
    )
    orgao_gerenciador: Mapped[Optional[str]] = mapped_column(
        String(300), nullable=True,
        comment="Órgão gerenciador da ARP (somente para modalidade ARP).",
    )
    tipo_instrumento: Mapped[Optional[str]] = mapped_column(
        String(20), nullable=True,
        comment="Tipo de instrumento: CONTRATO ou NOTA_EMPENHO (somente para modalidade CONTRATO).",
    )
    tipo_contratacao: Mapped[Optional[str]] = mapped_column(
        String(100), nullable=True,
        comment="Tipo de contratação (ex: Dispensa, Inexigibilidade, Pregão, etc.).",
    )


    # ── Vínculo com Projeto ─────────────────────────────────────────────────
    projeto_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("projetos.id", ondelete="RESTRICT"),
        nullable=False,
        comment="Projeto de origem (deve ter status 'Licitação concluída').",
    )

    # ── Dados da Contratação ────────────────────────────────────────────────
    fornecedor_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("fornecedores.id", ondelete="RESTRICT"),
        nullable=True,
        comment="Fornecedor vinculado ao contrato (FK para fornecedores).",
    )
    tipo_fornecedor_contrato: Mapped[Optional[str]] = mapped_column(
        String(30), nullable=True,
        comment="Tipo: REVENDEDOR, FABRICANTE ou REVENDEDOR_E_FABRICANTE.",
    )
    tipo_contrato: Mapped[TipoContratoEnum] = mapped_column(
        Enum(
            TipoContratoEnum,
            name="tipo_contrato_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
    )
    # ── Itens da Contratação ────────────────────────────────────────────────
    itens: Mapped[list["ItemContrato"]] = relationship(
        "ItemContrato",
        back_populates="contrato",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    # ── Vigência ────────────────────────────────────────────────────────────
    data_inicio_vigencia: Mapped[Optional[date]] = mapped_column(
        Date, nullable=True,
        comment="Data de início da vigência (pode diferir da assinatura).",
    )
    vigencia_meses: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True,
        comment="Quantidade de meses de vigência original (calculado automaticamente).",
    )
    prorrogacao_meses: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default="0",
        comment="Prorrogação total em meses acumulada via aditivos.",
    )
    data_assinatura: Mapped[date] = mapped_column(
        Date, nullable=False,
        comment="Data de assinatura do contrato.",
    )
    data_fim_vigencia: Mapped[date] = mapped_column(
        Date, nullable=False,
        comment="Data de encerramento da vigência.",
    )

    # ── Situação ────────────────────────────────────────────────────────────
    situacao_atual: Mapped[SituacaoContratoEnum] = mapped_column(
        Enum(
            SituacaoContratoEnum,
            name="situacao_contrato_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
        default=SituacaoContratoEnum.VIGENTE,
    )

    # ── Complexidade ────────────────────────────────────────────────────────
    complexidade: Mapped[ComplexidadeContratoEnum] = mapped_column(
        Enum(
            ComplexidadeContratoEnum,
            name="complexidade_contrato_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
        default=ComplexidadeContratoEnum.SIMPLES,
        server_default="simples",
        comment="Complexidade do contrato: simples, intermediaria ou complexa.",
    )


    # ── Timestamps ──────────────────────────────────────────────────────────
    criado_em: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    atualizado_em: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    # ── Relationships ───────────────────────────────────────────────────────
    projeto: Mapped["Projeto"] = relationship("Projeto", lazy="selectin")

    fornecedor_rel: Mapped[Optional["Fornecedor"]] = relationship(
        "Fornecedor", foreign_keys=[fornecedor_id], lazy="selectin",
    )

    # Equipe de Fiscalização (1:N via tabela contrato_equipe)
    equipe_membros: Mapped[list["ContratoEquipe"]] = relationship(
        "ContratoEquipe",
        back_populates="contrato",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    # Histórico / Auditoria (1:N)
    historico: Mapped[list["ContratoHistorico"]] = relationship(
        "ContratoHistorico",
        back_populates="contrato",
        order_by="desc(ContratoHistorico.data_hora)",
        lazy="selectin",
    )

    # Aditivos de Prazo (1:N)
    aditivos: Mapped[list["Aditivo"]] = relationship(
        "Aditivo",
        back_populates="contrato",
        cascade="all, delete-orphan",
        order_by="Aditivo.data_fim_vigencia.desc()",
        lazy="selectin",
    )

    # ── Constraints ─────────────────────────────────────────────────────────
    __table_args__ = (
        UniqueConstraint("numero", "ano", name="uq_contrato_numero_ano"),
    )

    @property
    def valor_total(self) -> Decimal:
        """Soma do valor total de todos os itens do contrato."""
        return sum((item.valor_total for item in self.itens), Decimal(0))

    def __repr__(self) -> str:
        return f"<Contrato #{self.id} {self.numero}/{self.ano} situacao={self.situacao_atual.value}>"


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  CONTRATO EQUIPE (Tabela para Titular + Substitutos)                   ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class ContratoEquipe(Base):
    """Membro da equipe de fiscalização de um contrato.
    Cada papel pode ter 1 Titular (is_titular=True) e N Substitutos.
    """
    __tablename__ = "contrato_equipe"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    contrato_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("contratos.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    servidor_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("servidores.id", ondelete="CASCADE"),
        nullable=False,
    )
    papel: Mapped[PapelEquipeEnum] = mapped_column(
        Enum(
            PapelEquipeEnum,
            name="papel_equipe_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
        comment="Papel do servidor na equipe.",
    )
    is_titular: Mapped[bool] = mapped_column(
        default=False,
        comment="True = Titular; False = Substituto.",
    )

    # ── Relationships ──────────────────────────────────────────────────────
    contrato: Mapped["Contrato"] = relationship("Contrato", back_populates="equipe_membros")
    servidor: Mapped["Servidor"] = relationship("Servidor", lazy="selectin")

    def __repr__(self) -> str:
        tipo = "Titular" if self.is_titular else "Substituto"
        return f"<ContratoEquipe #{self.id} papel={self.papel.value} {tipo}>"


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  ITEM DE CONTRATO (Múltiplos produtos/serviços)                        ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class ItemContrato(Base):
    __tablename__ = "item_contrato"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    contrato_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("contratos.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    catalogo_produto_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("catalogo_produtos.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Referência ao item padronizado do Catálogo de Produtos/Serviços.",
    )

    quantidade: Mapped[int] = mapped_column(
        Integer, nullable=False, default=1,
        comment="Quantidade de itens.",
    )
    valor_unitario: Mapped[Decimal] = mapped_column(
        Numeric(15, 2), nullable=False, default=0,
        comment="Valor unitário do item.",
    )

    # ── Catálogo Governamental (CATMAT ou CATSER) ──────────────────────────
    tipo_catalogo: Mapped[Optional[str]] = mapped_column(
        String(10), nullable=True,
        comment="Tipo de catálogo: 'CATMAT' (material) ou 'CATSER' (serviço).",
    )
    codigo_catalogo: Mapped[Optional[str]] = mapped_column(
        String(50), nullable=True,
        comment="Código numérico do catálogo (ComprasNet).",
    )

    # ── Vigência Granular ──────────────────────────────────────────────────
    data_inicio_vigencia: Mapped[Optional[date]] = mapped_column(
        Date, nullable=True,
        comment="Data de início do suporte/garantia específico do item.",
    )
    data_fim_vigencia: Mapped[Optional[date]] = mapped_column(
        Date, nullable=True,
        comment="Data de fim do suporte/garantia específico do item.",
    )

    # ── Relationships ──────────────────────────────────────────────────────
    contrato: Mapped["Contrato"] = relationship("Contrato", back_populates="itens")
    catalogo_produto: Mapped[Optional["CatalogoProduto"]] = relationship("CatalogoProduto")

    @property
    def valor_total(self) -> Decimal:
        return Decimal(self.quantidade) * self.valor_unitario

    def __repr__(self) -> str:
        return f"<ItemContrato #{self.id} qtd={self.quantidade} valor_unitario={self.valor_unitario}>"


# ╔══════════════════════════════════════════════════════════════════════════╗
# ║  CONTRATO HISTÓRICO (Log de Auditoria e Observações)                   ║
# ╚══════════════════════════════════════════════════════════════════════════╝


class TipoRegistroHistoricoEnum(str, enum.Enum):
    """Tipo de registro no histórico do contrato."""
    EDICAO_SISTEMA = "Edição de Sistema"
    OBSERVACAO_MANUAL = "Observação Manual"


class ContratoHistorico(Base):
    __tablename__ = "contrato_historico"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    contrato_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("contratos.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    data_hora: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False,
    )
    autor: Mapped[str] = mapped_column(
        String(200), nullable=False, default="Usuário do Sistema",
        comment="Autor da ação (temporário até JWT).",
    )
    tipo_registro: Mapped[TipoRegistroHistoricoEnum] = mapped_column(
        Enum(
            TipoRegistroHistoricoEnum,
            name="tipo_registro_historico_enum",
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
    )
    conteudo: Mapped[str] = mapped_column(
        Text, nullable=False,
        comment="Descrição da mudança ou texto da observação.",
    )

    # ── Relationships ──────────────────────────────────────────────────────
    contrato: Mapped["Contrato"] = relationship(
        "Contrato", back_populates="historico",
    )

    def __repr__(self) -> str:
        return f"<ContratoHistorico #{self.id} tipo={self.tipo_registro.value}>"


# ── Forward reference imports ──────────────────────────────────────────────
from app.models.projeto import Projeto, Servidor  # noqa: E402, F401
from app.models.fornecedor import Fornecedor  # noqa: E402, F401
from app.models.aditivo import Aditivo  # noqa: E402, F401
from app.models.catalogo import CatalogoProduto  # noqa: E402, F401
